File: app/src/main/java/com/demo/room/dao/PersonDao.java
package com.demo.room.dao;

import com.demo.room.bean.Person;

import java.util.List;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;
import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Single;

@Dao
public interface PersonDao {
    /**
     * 一次插入单条数据或多条
     * OnConflictStrategy.REPLACE:替换旧数据
     * OnConflictStrategy.ROLLBACK:回滚事务
     * OnConflictStrategy.ABORT:终止事务崩溃
     * OnConflictStrategy.FAIL:事务失败
     * OnConflictStrategy.IGNORE:忽略冲突
     */
    @Insert(onConflict = OnConflictStrategy.ABORT)
    Completable insert(Person... persons);

    @Query("DELETE FROM t_person")
    Completable deleteAll();

    /**
     * 一次删除单条数据或多条
     */
    @Delete
    Completable delete(Person... persons);

    /**
     * 一次更新单条数据或多条
     */
    @Update
    Completable update(Person... persons);

    @Query("SELECT * FROM t_person")
    Observable<List<Person>> getAll();

    @Query("SELECT * FROM t_person WHERE p_id= :pid")
    Single<Person> getPersonByPid(int pid);

    /**
     * 一次查找多个数据
     */
    @Query("SELECT * FROM t_person WHERE p_id IN (:pidList)")
    Single<List<Person>> getPersonListByPids(List<Integer> pidList);

    /**
     * 多个条件查找
     */
    @Query("SELECT * FROM t_person WHERE name = :name AND age = :age")
    Single<Person> getPersonByMultiCondition(String name, int age);
}

File: app/src/main/java/com/gxd/demo/room/dao/PersonDao.kt
package com.gxd.demo.room.dao

import androidx.room.Dao
import androidx.room.Query
import androidx.room.Transaction
import com.gxd.demo.room.data.PersonAndClothes
import com.gxd.demo.room.data.PersonClothes
import com.gxd.demo.room.data.PersonMinimal
import com.gxd.demo.room.table.Clothes
import com.gxd.demo.room.table.Person
import io.reactivex.Completable
import io.reactivex.Single

@Dao
interface PersonDao : BaseDao<Person> {
    @Query("DELETE FROM t_person")
    fun deleteAllRx(): Completable

    @Query("DELETE FROM t_person")
    fun deleteAll()

    @Query("SELECT * FROM t_person")
    fun queryAll(): Single<List<Person>>

    @Query("SELECT * FROM t_person WHERE p_id = :id")
    fun queryPersonById(id: Int): Single<Person>

    @Query("SELECT * FROM t_person WHERE p_id IN (:ids)")
    fun queryPersonListById(vararg ids: Int): Single<List<Person>>

    @Query("SELECT * FROM t_person WHERE name = :name AND age = :age")
    fun queryPerson(name: String, age: Int): Single<List<Person>>

    /**
     * 查询部分字段可以通过降低I/O开销提高查询速度
     */
    @Query("SELECT p_id,name,age FROM t_person")
    fun queryPersonMinimalList(): Single<List<PersonMinimal>>

    @Transaction
    fun updateAll(personList: List<Person>) {
        deleteAll()
        insert(personList)
    }

    @Query("SELECT * FROM t_person")
    fun queryPersonAndClothesList(): Single<List<PersonAndClothes>>

    @Query(
        "SELECT t_person.name AS personName, t_clothes.brand AS clothesBrand " +
                "FROM t_person,t_clothes WHERE t_person.p_id = t_clothes.owner_id"
    )
    fun queryIntermediateData(): Single<List<PersonClothes>>

    @Query("SELECT * FROM t_person JOIN t_clothes ON t_person.p_id = t_clothes.owner_id")
    fun queryMultimapData(): Single<Map<Person, List<Clothes>>>

//    @Query("SELECT * FROM fts_person WHERE fts_person MATCH :query")
//    fun search(query: String): Observable<List<PersonFTS>>
}